src/PrestaShopBundle/Entity/Repository/ShipmentRepository.php

```
@@ -55,4 +56,61 @@ public function save(Shipment $shipment): int
+    }
+
+    /**
+     * @param int $carrierId
+     * @param Shipment $shipmentToRemoveProduct
+     * @param ShipmentProduct[] $shipmentProductsToMove
+     */
+    public function splitShipment(int $carrierId, Shipment $shipmentToRemoveProduct, $shipmentProductsToMove): void
+    {
+        $shipmentProductsByOrderDetailId = array_reduce($shipmentToRemoveProduct->getProducts()->toArray(), function ($carry, ShipmentProduct $product) {
+            $carry[$product->getOrderDetailId()] = $product;
+
+            return $carry;
+        }, []);
+
+        $shipmentProducts = [];
+
+        $newShipment = new Shipment();
+        $newShipment->setCarrierId($carrierId);
+        $newShipment->setOrderId($shipmentToRemoveProduct->getOrderId());
+        $newShipment->setTrackingNumber(null);
+        $newShipment->setAddressId($shipmentToRemoveProduct->getAddressId());
+        $newShipment->setShippingCostTaxExcluded($shipmentToRemoveProduct->getShippingCostTaxExcluded());
+        $newShipment->setShippingCostTaxIncluded($shipmentToRemoveProduct->getShippingCostTaxIncluded());
+
+        foreach ($shipmentProductsToMove as $shipmentProductToMove) {
+            $orderDetailId = $shipmentProductToMove->getOrderDetailId();
+            $quantity = $shipmentProductToMove->getQuantity();
+            $existingProduct = $shipmentProductsByOrderDetailId[$orderDetailId] ?? null;
+            if ($existingProduct) {
+                $remainingQty = $existingProduct->getQuantity() - $quantity;
+                if ($remainingQty <= 0) {
+                    $shipmentToRemoveProduct->removeProduct($existingProduct);
+                } else {
+                    $existingProduct->setQuantity($remainingQty);
+                }
+            }
+            $shipmentProducts[] = (new ShipmentProduct())
+                ->setOrderDetailId($orderDetailId)
+                ->setQuantity($quantity)
+                ->setShipment($newShipment);
```

I'm wondering what case is covered by `$existingProduct === null`, is that a nominal case? Otherwise it may be better to throw an exception ?

Any product added in a shipment MUST BE deleted from another one before.